src/composed-graph.cpp

```diff
@@ -0,0 +1,156 @@
+}
+
+NicePathDecomposition perform_path_dec(Graph &g, std::vector<std::vector<vertex_t>> &U) {
+    auto pw_u = get_path_dec_U(g, U);
+    ComposedGraph CG(g, U, pw_u);
+    auto pw_comp = get_path_dec_comp(CG);
+    return CG.get_path_decomposition(pw_comp);
+}
+
+void ComposedGraph::check() {
+    int n = boost::num_vertices(_G);
+    _matchings.assign(n, -1);
+    _id_U.assign(n, -1);
+
+    for (int i = 0; i < _U.size(); ++i) {
+        for (vertex_t v : _U[i]) {
+            if (_id_U[v] < n) throw CorrectnessException(this, _id_U[v], i, v);
+            _id_U[v] = i;
+        }
+    }
+
+    for (vertex_t v = 0; v < n; ++v) {
+        if (_id_U[v] >= n) throw CorrectnessException(v, this);
+    }
+
+    for (auto it = boost::edges(_G).first; it != boost::edges(_G).second; ++it) {
+        edge_t e = *it;
+        vertex_t v = boost::source(e, _G);
+        vertex_t u = boost::target(e, _G);
+        if (_id_U[v] == _id_U[u]) continue;
+        if (_matchings[v] < n) throw CorrectnessException(v, _matchings[v], u, this);
+        if (_matchings[u] < n) throw CorrectnessException(u, _matchings[u], v, this);
+        _matchings[u] = v;
+        _matchings[v] = u;
+    }
+}
+
+ComposedGraph::CorrectnessException::CorrectnessException(vertex_t outer_vertex, ComposedGraph *G):
+_G(G), _outer_vertex(outer_vertex) {
+    _msg = "In ComposedGraph |_G| vertex |_outer_vertex| does not belongs to any U[i]";
+}
+
+ComposedGraph::CorrectnessException::CorrectnessException(ComposedGraph *G, int U_i, int U_j, vertex_t common_vertex):
+_G(G), _U_i(U_i), _U_j(U_j), _common_vertex(common_vertex) {
+    _msg = "|_G.U[_U_i]| and |_G.U[_U_j]| have common vertex |_common_vertex|";
+}
+
+ComposedGraph::CorrectnessException::CorrectnessException(vertex_t matching_vertex,
+                                  vertex_t adjacent_vertex_first,
+                                  vertex_t adjacent_vertex_second,
+                                  ComposedGraph *G):
+                                  _matching_vertex(matching_vertex),
+                                  _adjacent_vertex_first(adjacent_vertex_first),
+                                  _adjacent_vertex_second(adjacent_vertex_second) {
+    _msg = std::string("Vertex |_matching_vertex| has 2 outer edges ") +
+            std::string("(_matching_vertex, _adjacent_vertex_first) and ") +
+            std::string("(_matching_vertex, _adjacent_vertex_second)");
+}
+
+const char * ComposedGraph::CorrectnessException::what() const throw() {
+    return _msg.c_str();
+}
+
+NicePathDecomposition ComposedGraph::get_path_decomposition(NicePathDecomposition &pw_compressed) {
+    int n = boost::num_vertices(_G);
+    _pw_compressed = pw_compressed;
+
+    std::vector<bool> used(boost::num_vertices(_G), false);
+    std::vector<NicePathDecomposition::Bag> G_path_decomp;
+
+    for (auto bag : _pw_compressed.get_nice_bags()) {
+        if (bag.type != NicePathDecomposition::ADD_VERTEX) continue;
```

Comment:
What if we remove a node from the compressed path decomposition's bag? Shouldn't we remove some of the nodes in the corresponding block?

Reply:
No, they are instantly removed after adding node from compressed graph

Reply:
Can you explain it further? I am not sure where it happens.

Reply:
I am just copying all bags from small NicePathDecompositions so all vertices that was declared at small NicePathDecomposition was removed there

Reply:
So when do you remove _matchings[in_bag.vertex] nodes then?

Reply:
When set with _matchings[in_bag.vertex] appears